lib/kcerts/ssh.go

```diff
@@ -0,0 +1,198 @@
+package kcerts
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"encoding/json"
+	"fmt"
+	"github.com/enfabrica/enkit/lib/cache"
+	"github.com/mitchellh/go-homedir"
+	"golang.org/x/crypto/ssh"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const CAPrefix = "@cert-authority"
+const SSHDir = ".ssh"
+const KnownHosts = "known_hosts"
+
+const SSHCacheKey = "enkit_ssh_cache_key"
+
+// FindSSHDir will find the users ssh directory based on $HOME. If $HOME/.ssh does not exist
+// it will attempt to create it.
+func FindSSHDir() (string, error) {
+	hDir, err := homedir.Dir()
+	if err != nil {
+		return "", fmt.Errorf("could not find the home directory: %w", err)
+	}
+	sshDir := filepath.Join(hDir, SSHDir)
+	if err := os.Mkdir(sshDir, 0700); err != nil && !os.IsExist(err) {
+		return "", fmt.Errorf("could not create directory %s: %w", sshDir, err)
+	}
+	return sshDir, nil
+}
+
+// AddSSHCAToClient adds a public key to the $HOME/.ssh/known_hosts in the ssh-cert x509.1 format.
+// For each entry, it adds an additional line and does not concatenate
+func AddSSHCAToClient(publicKey ssh.PublicKey, hosts []string, sshDir string) error {
+	caPublic := string(ssh.MarshalAuthorizedKey(publicKey))
+	knownHosts := filepath.Join(sshDir, KnownHosts)
+	knownHostsFile, err := os.OpenFile(knownHosts, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0644)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return fmt.Errorf("could not create known_hosts file: %w", err)
+		}
+		return err
+	}
+	existingKnownHostsContent, err := ioutil.ReadAll(knownHostsFile)
+	if err != nil {
+		return fmt.Errorf("error reading %s: %w", knownHosts, err)
+	}
+	defer knownHostsFile.Close()
+	for _, dns := range hosts {
+		// caPublic terminates with a '\n', added by ssh.MarshalAuthorizedKey
+		publicFormat := fmt.Sprintf("%s %s %s", CAPrefix, dns, caPublic)
+		if strings.Contains(string(existingKnownHostsContent), publicFormat) {
+			continue
+		}
+		_, err = knownHostsFile.WriteString(publicFormat)
+		if err != nil {
+			return fmt.Errorf("could not add key %s to known_hosts file: %w", publicFormat, err)
+		}
+	}
+	return nil
+}
+
+type sshCache struct {
+	PID       int       `json:"pid"`
+	Sock      string    `json:"sock"`
+	TimeStamp time.Time `json:"time_stamp"`
+}
+
+// FindSSHAgent Will start the ssh agent in the interactive terminal if it isn't present already as an environment variable
+// Currently only outputs the env and does not persist it across terminals
+func FindSSHAgent(store cache.Store, ttl time.Duration) (string, int, error) {
+	envSSHSock := os.Getenv("SSH_AUTH_SOCK")
+	envSSHPID := os.Getenv("SSH_AGENT_PID")
+	if envSSHSock != "" && envSSHPID != "" {
+		fmt.Println("found in env")
+		pid, err := strconv.Atoi(envSSHPID)
```

Comment:
am ok fixing later, but add a TODO in case: the environment variables may be totally bogus.

Eg, `SSH_AGENT_PID` and `SSH_AUTH_SOCK` may be pointing to an `ssh-agent` that is long dead or long gone.
This happens more frequently than you think, pretty much every time you ssh on a machine with agent forwarding if you use `screen` and `tmux`: your screen session will reattach you to an old shell, likely pointing to an old agent / pid that no longer exists.

My take here: a) rather than return error, just ignore the environment `ssh-agent` if anything is fishy about it (eg, pid is not an integer), b) ideally, try to use the socket before assuming that the agent is available / alive.

Reply:
Added in todo, leaving this chain open in case you find anything else